src/sentry/integrations/messaging.py

```diff
@@ -0,0 +1,215 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+from django.urls import re_path
+from django.urls.resolvers import URLPattern
+from django.views.generic import View
+
+from sentry import analytics
+from sentry.incidents.action_handlers import ActionHandler, DefaultActionHandler
+from sentry.incidents.models.alert_rule import ActionHandlerFactory, AlertRuleTriggerAction
+from sentry.incidents.models.incident import Incident, IncidentStatus
+from sentry.integrations.base import IntegrationProvider
+from sentry.models.notificationaction import ActionService, ActionTarget
+from sentry.models.project import Project
+from sentry.rules import rules
+from sentry.rules.actions import IntegrationEventAction
+
+
+@dataclass(frozen=True)
+class MessagingIdentityLinkViewSet:
+    """An integration's set of view classes for linking and unlinking identities."""
+
+    link_personal_identity: type[View]
+    unlink_personal_identity: type[View]
+
+    # TODO: Allow these to be optional? Require for all messaging integrations?
+    link_team_identity: type[View] | None = None
+    unlink_team_identity: type[View] | None = None
+
+
+class MessagingIntegrationSpec(ABC):
+    """Represent the feature set for a messaging integration.
+
+    This class is intended to serve as a top-level "table of contents" for all the
+    code that supports integration of a third-party service with Sentry. The
+    IntegrationProvider class, which this class incorporates, provides specifications
+    for the base integration. This class's distinct purpose is to adds whatever
+    features and business logic are particular to messaging integrations.
+
+    This class is currently under development and does not yet represent the complete
+    feature set for messaging integrations. We should continue developing it by
+    adding anything that is required for all messaging integrations or common across
+    many of them, especially where there code or patterns are duplicated.
+    """
+
+    def initialize(self) -> None:
+        """Initialize a messaging integration.
+
+        We expect each MessagingIntegrationSpec implementation (i.e., concrete
+        subclass) to be instantiated once and to have this method called in an
+        `__init__` module, or somewhere else that is reliably evaluated on start-up.
+
+        See docstrings on `integration_provider` and
+        `get_identity_view_set_url_patterns` for things *not* covered by this method,
+        which require additional boilerplate.
+        """
+
+        AlertRuleTriggerAction.register_factory(_MessagingHandlerFactory(self))
+
+        if self.notify_service_action:
+            rules.add(self.notify_service_action)
+        if self.notification_sent:
+            analytics.register(self.notification_sent)
```

Comment:
Yes. However, since they are currently implemented for all messaging integrations, we could go ahead and make them required. I'm trying to remember my rationale here -- I think I intended it as a conceptual demonstration that they don't *have* to be required unless we want?

I should at least go ahead and make `notify_service_action` required -- the integration is pretty useless without it.